Specify Platform in ExecuteProcessResult output

gshuflin opened the pull request:
### Problem

We want to be able to track the actual Platform that an external process was ran on in as a field of a new type `FallibleExecuteProcessResultWithPlatform`. This is important for ensuring that some cross platform workflows (such as, a user running pants on a Mac remoting on a build farm with Linux build machines) work correctly. 

### Solution

This commit introduces the Rust boilerplate for creating `Platform` type and passing it across the FFI to Python. It is not yet complete - `cache.rs` has not yet been updated to correctly cache the Platform, and will incorrectly always return "Linux" until this is fixed in a follow-up PR. 

It also renames the string representation of the OSX Platform from "osx" to "darwin", to match what the Python code assumes.

### Result

There is now a new type `FallibleExecuteProcessResultWithPlatform` which has a new parameter `platform`, which will (incorrectly) always be the same as the local platform until a follow-up PR fixes this. The existing `FallibleExecuteProcessResult` and `ExecuteProcessResult` types work the same way as they previously did, but are now derived from `FallibleExecuteProcessResultWithPlatform` by dropping the platform information.

Eric-Arellano commented on src/rust/engine/src/context.rs:
```diff
@@ -180,7 +180,7 @@ impl Core {
             store.clone(),
             // TODO if we ever want to configure the remote platform to be something else we
             // need to take an option all the way down here and into the remote::CommandRunner struct.
```
Still relevant?

gshuflin replied:
For now, yes. This code does just assume that the remote platform is always Linux, and I think this is fine until the day comes when we want to support OSX remote build machines.

Eric-Arellano commented on src/python/pants/engine/isolated_process.py:
```diff
@@ -132,6 +131,19 @@ class FallibleExecuteProcessResult:
+
+
+@dataclass(frozen=True)
+class FallibleExecuteProcessResultWithPlatform:
```
Do you think that we want `ExecuteProcessResultWithPlatform` so that rules can request that as an alternative to `FallibleExecuteProcessResultWithPlatform`?

gshuflin replied:
Possibly, but we can add that in another PR when we actually need it, this one is long enough already

Eric-Arellano commented on src/python/pants/engine/isolated_process.py:
```diff
@@ -132,6 +131,19 @@ class FallibleExecuteProcessResult:
+
+
+@dataclass(frozen=True)
+class FallibleExecuteProcessResultWithPlatform:
```
While this is quite wordy, this name fits well with conventions we've been using in V2 for when you want to have some extra information attached to a normal type, e.g. `AddressWithOrigin` and `HydratedTargetWithOrigin`.

gshuflin replied:
Yeah it is wordy, but I don't have a better name in mind.

illicitonion commented on src/rust/engine/process_execution/src/cache.rs:
```diff
@@ -100,6 +102,7 @@ impl CommandRunner {
             execute_response,
             vec![],
             context.workunit_store,
+            Platform::Linux, //TODO this is incorrect and will be fixed in a follow-up PR
```
Just to check, can you sketch out how you're planning to fix it before merging?

gshuflin replied:
https://github.com/pantsbuild/pants/pull/9198 is the draft PR that fixes this, that I need to give a little bit more attention to before flipping it out of draft mode.

illicitonion commented on src/rust/engine/process_execution/src/cache_tests.rs:
```diff
@@ -98,6 +99,7 @@ fn run_roundtrip(script_exit_code: i8) -> RoundtripResults {
 }
 
 #[test]
+#[ignore] //TODO(gshuflin) this test will fail on a mac until a follow-up removes the hard-coding of "Linux" in the cache.
```
If you default your placeholder to `Platform::current().unwrap()` rather than `Platform::Linux`, this test will keep working in the mean time, while being equally wrong in production :)

gshuflin replied:
@illicitonion doing that breaks a bunch of other tests; cf. the below comment I'm just gonna rebase the PR that actually finishes implementing the feature properly on top of these commits, even if it makes for a long PR.

illicitonion commented on src/rust/engine/process_execution/src/lib.rs:
```diff
@@ -280,10 +307,11 @@ pub struct ExecuteProcessRequestMetadata {
 ///
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub struct FallibleExecuteProcessResult {
+pub struct FallibleExecuteProcessResultWithPlatformWithPlatform {
```
Why the double `WithPlatform`?

gshuflin replied:
typo! will fix

gshuflin commented:
> > ExecuteProcessResult will now have a new parameter platform, which will (incorrectly) always be Linux until a follow-up PR fixes this.
> 
> I don' think this is true. The local command runner injects the local platform string into requests that it runs. What am I missing? Do you mean that all remotely run commands will have linux hardcoded?

Ah, this is no longer true, since I changed this in response to a comment Daniel made earlier (and I will update the PR description). Remotely run commands aren't currently being cached with the correct Platform (see https://github.com/pantsbuild/pants/pull/9198 which fixes that). Originally I hard-coded the Platform that the cache outputs to always be Linux (until that PR actually records the platform a request ran on), but on Daniel's suggestion I changed it to just always output the local platform. This is still incorrect, but it avoids temporarily breaking a test on macs until I can get that PR out of draft mode, reviewed and committed.

gshuflin commented:
So it turns out that without https://github.com/pantsbuild/pants/pull/9198 implement it's *really* hard to get tests to pass on osx; and I think my time from here on out is better spent just finishing that PR (and asking for a review on that work) than on trying to get this PR to pass tests and become mergeable. I will rebase 9198 on top of these commits, count these commits as already reviewed, and ask for a new review on 9198 only on the new commits there.

hrfuller commented:
Feel free to add me.
